tests/test_options.py

```diff
@@ -96,3 +96,54 @@ def test_options_prefix():
     options = petsctools.OptionsManager({}, options_prefix="myobj",
                                         default_prefix="firedrake")
     assert options.options_prefix.startswith("myobj_")
+
+
+@pytest.mark.skipnopetsc4py
+def test_default_options():
+    from petsc4py import PETSc
+
+    parameters = {
+        'options_left': 0,
+        'unrelated_option': 0,
+        'base_opt1': 1,
+        'base_opt2': 2,
+        'base_opt3': 3,
+        'base_0_opt3': 4,
+        'base_1_opt3': 5,
+        'base_2_opt4': 6,
+    }
+    options = PETSc.Options()
+    for k, v in parameters.items():
+        options[k] = v
+
+    # default_options = {"opt1": 1, "opt2": 2, "opt3": 3}
+    default_options = petsctools.get_default_options(
+        base_prefix="base", custom_prefix_endings=("0", "1", "2"))
+
+    assert len(default_options) == 3
+    assert default_options["opt1"] == "1"
+    assert default_options["opt2"] == "2"
+    assert default_options["opt3"] == "3"
+
+    options0 = petsctools.OptionsManager(
+        parameters=default_options, options_prefix="base_0")
```

Comment:
I find the way that this is used in practice to be quite confusing. `get_default_options` is sort of a filter that you apply to the global options before passing to the `OptionsManager`? I wonder if you could instead have something like:
```py
options0 = petsctools.OptionsManager(
        parameters=parameters, options_prefix="base_0", shared_defaults_prefix="base")
```
that would do this filtering internally instead.

Reply:
What if you had a context manager that added suffix-ed defaults to the available options? Something like
```py
with outer_options.inserted_options():
    # so PETSc.Options is 
    # {'base_opt0': 0,
	#	'base_opt1': 1,
	#	'base_0_opt0': 2,
	#	'base_1_opt2': 3}
    with default_options(base_prefix="base", custom_prefix_endings=('0', '1', '2'):
        # so PETSc.Options is 
        # { 'base_0_opt0': 2,
        #   'base_0_opt1': 1,
        #   'base_1_opt0': 0,
        #   'base_1_opt1': 1,
        #   'base_1_opt2': 3,
        #   'base_2_opt0': 0,
        #   'base_2_opt1': 1}
	    inner_options_0 = OptionsManager(options_prefix='base_0')
        # etc
```

Reply:
Oh boy. Maybe? I'll have to have a think about that.

Reply:
I made it an actual class with a `property` for each attribute so it can make sure that the prefixes are actually valid (string ending in underscore), but otherwise it does basically what I wrote above.